application/core/plugins/AzureOAuthSMTP/AzureOAuthSMTP.php

```diff
@@ -0,0 +1,295 @@
+use PHPMailer\PHPMailer\PHPMailer;
+
+class AzureOAuthSMTP extends SmtpOAuthPluginBase
+{
+    protected $storage = 'DbStorage';
+    protected static $description = 'Core: Adds Azure OAuth support for email sending';
+    protected static $name = 'AzureOAuthSMTP';
+
+    /** @inheritdoc, this plugin doesn't have any public method */
+    public $allowedPublicMethods = [];
+
+    /** @inheritdoc */
+    protected $credentialAttributes = ['clientId', 'clientSecret', 'tenantId'];
+
+    protected $settings = [
+        'help' => [
+            'type' => 'info',
+            'content' => '',
+        ],
+        'clientId' => [
+            'type' => 'string',
+            'label' => 'Client ID',
+        ],
+        'clientSecret' => [
+            'type' => 'string',
+            'label' => 'Client Secret',
+        ],
+        'tenantId' => [
+            'type' => 'string',
+            'label' => 'Tenant ID',
+        ],
+    ];
+
+    public function init()
+    {
+        // This plugin is only compatible with PHP 7.3 and above,
+        // while LimeSurvey is still compatible with PHP 7.2, so
+        // we need to make sure we only load composer's autoload
+        // on PHP 7.3 and above. We also need to prevent activation
+        // of the plugin on PHP 7.2 and below.
+        // TODO: Remove this once we drop support for PHP 7.2
+        $this->subscribe('beforeActivate');
+        if (!(PHP_VERSION_ID >= 70300)) {
+            return;
+        }
+        require_once __DIR__ . '/vendor/autoload.php';
+
+        $this->subscribe('listEmailPlugins');
+        $this->subscribe('afterSelectEmailPlugin');
+        $this->subscribe('MailerConstruct');    // Handler defined in SmtpOAuthPluginBase
+        $this->subscribe('beforePrepareRedirectToAuthPage');
+        $this->subscribe('beforeRedirectToAuthPage');   // Handler defined in SmtpOAuthPluginBase
+        $this->subscribe('afterReceiveOAuthResponse');  // Handler defined in SmtpOAuthPluginBase
+
+        $this->subscribe('beforeEmailDispatch');
+    }
+
+    /**
+     * TODO: Remove this once we drop support for PHP 7.2
+     */
+    public function beforeActivate()
+    {
+        if (!(PHP_VERSION_ID >= 70300)) {
+            $event = $this->getEvent();
+            $event->set('success', false);
+            $event->set('message', gT("This plugin requires PHP 7.3 or above."));
+        }
+    }
+
+    /**
+     * @inheritdoc
+     * Update the information content
+     */
+    public function getPluginSettings($getValues = true)
+    {
+        $settings = parent::getPluginSettings($getValues);
+        $settings['help']['content'] = $this->getHelpContent();
+        $settings['clientId']['label'] = gT("Client ID");
+        $settings['clientSecret']['label'] = gT("Client Secret");
+        $settings['tenantId']['label'] = gT("Tenant ID");
+
+        return $settings;
+    }
+
+    private function getHelpContent()
+    {
+        $this->subscribe('getPluginTwigPath');
```

Comment:
Yes it is :)

Reply:
Where is it documented? I didn't find it on the plugin event manual page.

Reply:
Found it. I don't understand why it's needed tho, can't you just use `$this->renderPartial()` and it will look in the views/ folder inside the plugin automatically?

Reply:
> Found it. I don't understand why it's needed tho, can't you just use `$this->renderPartial()` and it will look in the views/ folder inside the plugin automatically?

Currently no. View folder are not added automatically.

Reply:
> > Found it. I don't understand why it's needed tho, can't you just use `$this->renderPartial()` and it will look in the views/ folder inside the plugin automatically?
> 
> Currently no. View folder are not added automatically.


        $content = $this->api->renderTwig(__DIR__ . '/views/index.twig', $data);

https://github.com/olleharstedt/MassAction/blob/master/MassAction.php#L228

That's not enough?

Reply:
> > > Found it. I don't understand why it's needed tho, can't you just use `$this->renderPartial()` and it will look in the views/ folder inside the plugin automatically?
> > 
> > 
> > Currently no. View folder are not added automatically.
> 
> ```
>     $content = $this->api->renderTwig(__DIR__ . '/views/index.twig', $data);
> ```
> 
> https://github.com/olleharstedt/MassAction/blob/master/MassAction.php#L228
> 
> That's not enough?

Oh !! You're right !

getPluginTwigPath isb only needed for Public page : to allow theme to replace the twig file !

Sorry !

Can be removed : you're right